File: PRACTICAS/CAMBIAR_RUBY/RUBY2/Civitas/lib/vista_textual.rb
#encoding:utf-8
require_relative 'respuestas'
require_relative 'salidas_carcel'
require 'io/console'


module Civitas

  class Vista_textual
    def initialize
      @juego_model
      @i_gestion = -1
      @i_propiedad = -1
      @separador = "======================"
    end

      attr_reader :juego_model
    
    def mostrar_estado(estado)
      puts estado
    end

    
    def pausa
      print "Pulsa una tecla"
      STDIN.getch
      print "\n"
    end

    def lee_entero(max,msg1,msg2)
      ok = false
      begin
        print msg1
        cadena = gets.chomp
        begin
          if (cadena =~ /\A\d+\Z/)
            numero = cadena.to_i
            ok = true
          else
            raise IOError
          end
        rescue IOError
          puts msg2
        end
        if (ok)
          if (numero >= max)
            ok = false
          end
        end
      end while (!ok)

      return numero
    end



    def menu(titulo,lista)
      tab = "  "
      puts titulo
      index = 0
      lista.each { |l|
        puts tab+index.to_s+"-"+l
        index += 1
      }

      opcion = lee_entero(lista.length,
                          "\n"+tab+"Elige una opcion: ",
                          tab+"Valor erroneo")
      return opcion
    end

    def salir_carcel

      lista_salidas_carcel = [SalidasCarcel::PAGANDO, SalidasCarcel::TIRANDO]
      opcion = menu("Elige la forma para intentar salir de la carcel", ["Pagando","Tirando el dado"])


      return lista_salidas_carcel[opcion]
    end
    def comprar
      lista_respuestas=[Respuestas::NO,Respuestas::SI]
      opcion = menu("¿Deseas comprar la calle a la que has llegado?",["SI","NO"])


      return lista_respuestas[opcion]
    end
    
    def gestionar
      opcion = menu("¿Que gestion inmobiliaria desea hacer?",
        ["VENDER","HIPOTECAR","CANCELAR HIPOTECA","CONSTRUIR CASA","CONSTRUIR HOTEL","CONSTRUIR IMPERIO", "TERMINAR"])

      @i_gestion = opcion

      if(@i_gestion != 5)
        tab = " "
        puts "¿Sobre que propiedad desea realiza la gestion? [0,#{@juegoModel.get_jugador_actual.propiedades.length-1}]"
        @i_propiedad = lee_entero(@juegoModel.get_jugador_actual.propiedades.length,
        "\n"+tab+"Elige la propiedad sobre la que aplicar la gestion: ", tab+"valor erroneo")
      end
    end

    def get_gestion
      return @i_gestion
    end

    def get_propiedad
      return @i_propiedad
    end

    def mostrar_siguiente_operacion(operacion)
      puts "La siguiente operacion que se va a realizar es: #{operacion}"
    end

    def mostrar_eventos
      puts "\nLos eventos pendientes son: "
      while(Diario.instance.eventos_pendientes)

      puts Diario.instance.leer_evento
      end
    end

    def set_civitas_juego(civitas)
         @juego_model=civitas
         self.actualizar_vista
    end

    def actualizar_vista
      puts "#{@juego_model.info_jugador_texto}"
    end
  end
end

File: PRACTICAS/practica3/Civitas_P1/src/JuegoTexto/VistaTextual.java
package JuegoTexto;

import civitas_p1.CivitasJuego;
import civitas_p1.Diario;
import civitas_p1.OperacionesJuego;
import civitas_p1.SalidasCarcel;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Scanner;
import civitas_p1.Respuestas;

public class VistaTextual {
  
  CivitasJuego juegoModel;
  int iGestion=-1;
  int iPropiedad=-1;
  private static String separador = "=====================";
  
  private Scanner in;
  
  public VistaTextual () {
    in = new Scanner (System.in);
  }
  
  void mostrarEstado(String estado) {
    System.out.println (estado);
  }
              
  void pausa() {
    System.out.print ("Pulsa una tecla: ");
    in.nextLine();
  }

  int leeEntero (int max, String msg1, String msg2) {
    Boolean ok;
    String cadena;
    int numero = -1;
    do {
      System.out.print (msg1);
      cadena = in.nextLine();
      try {  
        numero = Integer.parseInt(cadena);
        ok = true;
      } catch (NumberFormatException e) { // No se ha introducido un entero
        System.out.println (msg2);
        ok = false;  
      }
      if (ok && (numero < 0 || numero >= max)) {
        System.out.println (msg2);
        ok = false;
      }
    } while (!ok);

    return numero;
  }

  int menu (String titulo, ArrayList<String> lista) {
    String tab = "  ";
    int opcion;
    System.out.println (titulo);
    for (int i = 0; i < lista.size(); i++) {
      System.out.println (tab+i+"-"+lista.get(i));
    }

    opcion = leeEntero(lista.size(),
                          "\n"+tab+"Elige una opción: ",
                          tab+"Valor erróneo");
    return opcion;
  }

  SalidasCarcel salirCarcel() {
    int opcion = menu ("Elige la forma para intentar salir de la carcel",
      new ArrayList<> (Arrays.asList("Pagando","Tirando el dado")));
    return (SalidasCarcel.values()[opcion]);
  }

    Respuestas comprar() {
  
    int opcion = menu ("¿Deseas comprar la calle a la que has llegado?",
        new ArrayList<> (Arrays.asList("NO","SI")));
    return (Respuestas.values()[opcion]);
  }

  void gestionar () {

    String cabecera = "¿Que gestion desea realizar?";

    ArrayList<String> opciones = new ArrayList<>();
    opciones.add("VENDER");
    opciones.add("HIPOTECAR");
    opciones.add("CANCELAR_HIPOTECA");
    opciones.add("CONSTRUIR_CASA");
    opciones.add("CONSTRUIR_HOTEL");
    opciones.add("TERMINAR");

    iGestion = menu(cabecera, opciones);

    if (iGestion != 5){

        if (juegoModel.getJugadorActual().getPropiedades().size() <= 0){
            System.out.println("\nOperacion no Permitida, no tienes Propiedades.\n");
            gestionar();
        }
        else{
            Scanner sc = new Scanner(System.in);
            System.out.println("\nTienes " + juegoModel.getJugadorActual().getPropiedades().size() + " propiedades.\n ");
            ArrayList<String> properties = new ArrayList<>();
            for(int i = 0; i < juegoModel.getJugadorActual().getPropiedades().size(); i++){

                properties.add(juegoModel.getJugadorActual().getPropiedades().get(i).getNombre());

            }
            iPropiedad = menu ("Elige la propiedad sobre la que quieres realizar la opeacion: ", properties);
            while (iPropiedad < 0 || iPropiedad > juegoModel.getJugadorActual().getPropiedades().size()){
                System.out.println("\nHay " + juegoModel.getJugadorActual().getPropiedades().size() + " propiedades. \n");
                iPropiedad = sc.nextInt();
                System.out.println("\nValor de npropiedad " + iPropiedad + ". \n");
            }       
            
        }

    }


  }
  /*
  void gestionar () {

    String cabecera = "¿Que gestion desea realizar?";

    ArrayList<String> opciones = new ArrayList<>();
    opciones.add("VENDER");
    opciones.add("HIPOTECAR");
    opciones.add("CANCELAR_HIPOTECA");
    opciones.add("CONSTRUIR_CASA");
    opciones.add("CONSTRUIR_HOTEL");
    opciones.add("TERMINAR");

    iGestion = menu(cabecera, opciones);

    if(iGestion != 5){

        ArrayList<String> properties = new ArrayList<>();

        for(int i = 0; i < juegoModel.getJugadorActual().getPropiedades().size(); i++){

            properties.add(juegoModel.getJugadorActual().getPropiedades().get(i).getNombre());

        }
        if(!properties.isEmpty())
            iPropiedad = menu ("Elige la propiedad sobre la que quieres realizar la opeacion: ", properties);
        else
            System.out.println("No tienes propiedades, pulsa 5 para terminar.");
    }


  }
  
  
    String cabecera = "¿Que gestion desea realizar?";
    
    ArrayList<String> opciones = new ArrayList<>();
    opciones.add("VENDER");
    opciones.add("HIPOTECAR");
    opciones.add("CANCELAR_HIPOTECA");
    opciones.add("CONSTRUIR_CASA");
    opciones.add("CONSTRUIR_HOTE");
    opciones.add("TERMINAR");

    int elegido = menu(cabecera, opciones);
    
    switch (elegido){
        case 0: iGestion = 0;
        break;
        case 1: iGestion = 1;
        break;
        case 2: iGestion = 2;
        break;
        case 3: iGestion = 3;
        break;
        case 4: iGestion = 4;
        break;
        case 5: iGestion = 5;
        break;
    }
    
    if (iGestion != 5){
        Scanner sc = new Scanner(System.in);
        System.out.println("Elige el indice de la propiedad que quieres gestionar: ");
        int npropiedad = sc.nextInt();
        while (npropiedad < 0 || npropiedad > juegoModel.getJugadorActual().getPropiedades().size()){
            System.out.println("\nHay " + juegoModel.getJugadorActual().getPropiedades().size() + " propiedades. \n");
            npropiedad = sc.nextInt();
            System.out.println("\nValor de npropiedad " + npropiedad + ". \n");
        }
        iPropiedad = npropiedad;
    }
    
    
  }


  
  void gestionar () {

      int opcion = menu("¿Que gestión inmobiliaria desea hacer?",
        new ArrayList<> (Arrays.asList("VENDER","HIPOTECAR","CANCELAR HIPOTECA",
        "CONSTRUIR CASA","CONSTRUIR HOTEL", "TERMINAR")));

      iGestion = opcion;

      if(juegoModel.getJugadorActual().tieneAlgoQueGestionar() && opcion!=5){

          int indice = menu ("¿Sobre qué propiedad desea realiza la gestion?",
          new ArrayList<>(Arrays.asList("0 - "+juegoModel.getJugadorActual().getPropiedades().size())));

          iPropiedad = indice;

      }else{
          System.out.println("El jugador actual no tiene nada que gestionar");
      }


  }

*/
  
  public int getGestion(){return iGestion;}
  
  public int getPropiedad(){return iPropiedad;}
    

  void mostrarSiguienteOperacion(OperacionesJuego operacion) {
  
  System.out.println ("La siguiente operacion que se va a realizar es: " + operacion.name());
      
  }


  void mostrarEventos() {  
      
  System.out.println("\nLos eventos pendientes son: ");
  while(Diario.instance.eventosPendientes()){
      
      System.out.println(Diario.instance.leerEvento());
      
  }
         
  }
  
  public void setCivitasJuego(CivitasJuego civitas){ 
        juegoModel=civitas;
        this.actualizarVista();

    }
  
  void actualizarVista(){         
      
      System.out.println(juegoModel.infoJugadorTexto());
      
  } 
}
